### ISSUE TITLE
Benchmark, improve netdx efficiency for parallel, big network sims

### ISSUE THREAD

**Issue by smjenness** [statnet/EpiModel#490]
It seems like a lot more computational time is spent on merging and post-processing the data rather than simulating the networks here (a ratio of 5x-10x for a 25k node network):

```r
        cluster.size <- min(nsims, ncores)
        registerDoParallel(cluster.size)

        diag.sim <- foreach(i = 1:nsims) %dopar% {
          fit.sim <- simulate(fit, basis = fit$newnetwork,
                              control = set.control.ergm)
          simulate(fit.sim,
                   formation = formation,
                   dissolution = dissolution,
                   coef.form = coef.form,
                   coef.diss = coef.diss$coef.crude,
                   time.slices = nsteps,
                   constraints = constraints,
                   monitor = nwstats.formula,
                   nsim = 1,
                   control = set.control.stergm)
        }
```

Check with benchmarking more formally, and see what improvements are possible.

**Comment by AdrienLeGuillou** [comment]
I benchmarked the code above with this test:

```r
pkgload::load_all()

n <- 25 * 1e3

nw <- network_initialize(n = n)
est <- netest(nw, formation = ~edges, target.stats = n / 2,
  coef.diss = dissolution_coefs(~offset(edges), 10, 0),
  edapprox = FALSE, verbose = FALSE)


profvis::profvis({
  dx <- netdx(est, nsims = 8, ncores = 8, nsteps = 100, 
              verbose = FALSE, dynamic = TRUE)
})
```

I also tried without parallelization.

When `ncores` is set to 8, the post processing, especially this call to `as.data.frame` https://github.com/EpiModel/EpiModel/blob/main/R/netdx.R#L294 takes 17 secondes out of the 24 total.

When `ncores` is set to 1, this same call still takes 17 seconds but the simulation being run sequentially take 100 seconds.

This is because `as.data.frame` is always run after the simulations are done with `lapply` so always sequentially.

A solution to speed this up would be to refactor `netdx` with a function to make the diagnostics for 1 simulation, model simulation and post processing and then either run it multiple times sequentially or in parallel. This would be similar to what is now in `netsim` with the `netsim_loop` function doing the simulation for each replication and `netsim` handling whether the replications are run sequentially or in parallel.

**Comment by chad-klumb** [comment]
Have you considered getting `changes` back from `simulate`, instead of a `networkDynamic`?  This should eliminate the need to construct the `networkDynamic`, and also to call `as.data.frame` on it, since `changes` (plus the initial `edgelist`) contain the essential information in that data frame.  Some of the formation/dissolution stat crunching could probably be done in parallel as well, with a final (relatively "small") combination of information across processes occurring at the end.

As for `networkDynamic` vs. `changes`, here is a simple example similar to Adrien's:

```
require(tergm)
nw <- network.initialize(25000, dir = FALSE)

logit <- function(p) log(p/(1 - p))
D <- 10

t0d <- Sys.time()
nwd <- simulate(nw ~ Form(~edges) + Persist(~edges),
                coef = c(logit(1/25000) - log(D), log(D - 1)),
                time.slices = 100,
                dynamic = TRUE)
t1d <- Sys.time()
df <- as.data.frame(nwd)
t2d <- Sys.time()

t0c <- Sys.time()
ch <- simulate(nw ~ Form(~edges) + Persist(~edges),
               coef = c(logit(1/25000) - log(D), log(D - 1)),
               output = "changes",
               time.slices = 100,
               dynamic = TRUE)
t1c <- Sys.time()

print(t1d - t0d)
print(t2d - t1d)
print(t1c - t0c)
```

which gives

```
> print(t1d - t0d)
Time difference of 13.85116 secs
> print(t2d - t1d)
Time difference of 2.400375 secs
> print(t1c - t0c)
Time difference of 2.21475 secs
```

So getting `changes` alone is ~7x faster than getting a `networkDynamic` and calling `as.data.frame` in this particular (toy) example.  The data frame has some extra columns that `changes` does not come with, but the information in those columns can be extracted from `changes` (and the initial `edgelist`) in what should be substantially less time than it takes to construct an intermediate `networkDynamic`.

**Comment by smjenness** [comment]
@chad-klumb : Interesting. That seems like a good solution here. Once @sgoodreau is complete with the heterogenous model diagnostics coding, perhaps we can discuss what this revision would look like.

**Comment by sgoodreau** [comment]
Yes, let me wrap up the hetergeneous netdx code first so we don't get lots of conflicts.  Should be within a week.

**Comment by chad-klumb** [comment]
Sounds good to me.
